cql3: remove linearizations in the write path

michoecho opened the pull request:
As a part of the effort of removing big, contiguous buffers from the codebase,
cql3::raw_value should be made fragmented. Unfortunately a straightforward
rewrite to a fragmented buffer type is not possible, because we want
cql3::raw_value to be compatible with cql3::raw_value_view, and we want that
view to be based on fragmented_temporary_buffer::view, so that it can be
used to view data coming directly from seastar without copying.

This patch makes cql3::raw_value fragmented by making cql3::raw_value_view
a `variant` of managed_bytes_view and fragmented_temporary_buffer::view.

Code users which depended on `cql3::raw_value` being `bytes`,
and cql::raw_value_view being `fragmented_temporary_buffer::view` underneath
were adjusted to the new, dual representation, mainly through the
`cql3::raw_value_view::with_value` visitor and deserialization/validation
helpers added to `cql3::raw_value_view`.

The second part of this series gets rid of linearizations occuring when processing
compound types in the CQL layer. This is achieved by storing their elements in
`managed_bytes` instead of `bytes` in the partially deserialized form (`lists::value`
`tuples::value`, etc.) outputting `managed_bytes` instead of `bytes` in functions
which go from the partially deserialized form to the atomic cell format (for frozen
types), and avoiding calling deserialize/serialize on individual elements when
it's not necessary. (It's only necessary for CQLv2, because since CQLv3 the format
on the wire is the same as our internal one).

The above also forces some changes to `expression.cc`, and `restrictions`, mainly because
`IN` clauses store their arguments as `lists` and `tuples`, and the code which handled
this clause expected `bytes`.

After this series, the path from prepared CQL statements to `atomic_cell_or_collection`
is almost completely linearization-free. The last remaining place is `collection_mutation_description`,
where map keys are linearized to `bytes`.

michoecho commented:
Technically it's mergeable, but I'm marking it as a draft, because I'm not sure if it's a step in the right direction, and the PR probably isn't very useful on it's own anyway. I'd rather try to comb out all linearizations from the entire `cql3/` in one bigger series. (Note that `raw_value` still has a `bytes` constructor. This PR adjusts the users of `raw_value` (so that they do not depend on the internal representation of `raw_value`) but not the producers yet. It won't help much if `raw_value` is `managed_bytes` inside if it's constructed from `bytes`.)

@avikivity @denesb Please take a look. Is `variant` a reasonable way to go about this?

denesb commented:
I have a slight preference towards a polymorphic view which in turn is a `FragmentedView` because I think it would be cleaner, but I want to hear more opinions as I'm not too familiar with the cql code.

avikivity commented on test/boost/transport_test.cc:
```diff
@@ -26,6 +26,22 @@
+    if (a.is_value()) {
+        return b.is_value() && b.with_value([&] (const FragmentedView auto& v2) {
+            return a.with_value([&] (const FragmentedView auto& v1) {
+                return equal_unsigned(v1, v2);
```
This will be compiled 4 times!

avikivity commented:
> Technically it's mergeable, but I'm marking it as a draft, because I'm not sure if it's a step in the right direction, and the PR probably isn't very useful on it's own anyway. 

It's good practice to merge things early, to avoid having to resolve conflicts later and to let the code soak in tests incrementally.

> I'd rather try to comb out all linearizations from the entire `cql3/` in one bigger series.

I'm perfectly happy to merge things piecemeal. It reduces the effort to review and less work needs to be thrown away if there is a mistake in the direction (at least, a mistake that is detected during review).

> (Note that `raw_value` still has a `bytes` constructor. This PR adjusts the users of `raw_value` (so that they do not depend on the internal representation of `raw_value`) but not the producers yet. It won't help much if `raw_value` is `managed_bytes` inside if it's constructed from `bytes`.)
> 
> @avikivity @denesb Please take a look. Is `variant` a reasonable way to go about this?


I think it is. First, it's there already, so it's the path of least resistance. I'd prefer a non-polymorphic view, but I don't know if it's doable, and in any case it's secondary to changing the APIs. Once the APIs are changed we can play with the internal representation.

avikivity commented:
> I have a slight preference towards a polymorphic view which in turn is a `FragmentedView` because I think it would be cleaner, but I want to hear more opinions as I'm not too familiar with the cql code.

A polymorphic view will be slower. In every operation involving the view, multiple virtual methods have to be called. With the variant, there is one virtual method call that dispatches to to generated code that is bound to the concrete view behind it, and so can be inlined.

avikivity commented on cql3/values.hh:
```diff
@@ -54,30 +55,39 @@ struct raw_value_view {
         : _data{data}
     {}
+    raw_value_view(managed_bytes_view data)
+        : _data{data}
```
A possible optimization is to add bytes_view to the variant, and here (and in the temporary constructor) select the bytes_view branch if there are <= 1 fragments. This will bloat the generated code more (we'll have 9 versions of the equal() function), but create shorter paths for the very common case of a value fitting in a single fragment.

Let's defer it so it can come with some measurement that proves it helps.

michoecho replied:
What do you mean by "shorter paths" in this case?

If anything, I could see the benefit of increased branch predictability (because with_value would almost always choose the same branch), and decreased icache usage (because the code paths for fragmented_temporary_view::view and managed_bytes_view would be almost completely cold).
But:
1. Branch predictability and icache impact would probably need specialized microbenchmarks to measure (because our existing ones are predictable in the first place).
2. The entire thing sounds like a "100 CPU cycles per query" type of deal.

Probably not worth it.

avikivity replied:
Shorter paths means that we make a decision once in a higher level, then work with smaller and simpler data structures below.

There are three classes of workloads:
1. One which has mostly small values. Here, each cycle is important, because all of the work is spent on making decisions and the per-byte work is negligible.
2. One which has mostly large values. Here, there's no point in counting cycles since a lot of the work is moving large amounts of data around (we also expect higher miss rates from the row cache).
3. For completeness, a mixed workload.

My goal is to optimize for the small-value case since it's the one that's going to be CPU bound. The large-value and mixed-value workloads are going to be dominated by data movement (across the network and within the core). So branch prediction and instruction counts matter mostly for the small-value workload. So for the workload that benefits, it will predict the single-fragment case well.

It's not 25 cycles per query, but 25 cycles each time we touch a value, which can be dozens of times per query. It is still small compared to overall query cost (which can be 60,000+ cycles), but not negligible.

avikivity commented on cql3/values.hh:
```diff
@@ -86,38 +96,67 @@ public:
+    decltype(auto) with_value(Func f) const {
+        switch (_data.index()) {
+        case 0: return f(std::get<fragmented_temporary_buffer::view>(_data));
+        default: return f(std::get<managed_bytes_view>(_data));
```
Probably slightly faster to use std::get<const managed_bytes_view&>() to avoid copying the view (cheap as it is).

michoecho replied:
std::get returns a reference, not a copy.

denesb commented:
> > I have a slight preference towards a polymorphic view which in turn is a `FragmentedView` because I think it would be cleaner, but I want to hear more opinions as I'm not too familiar with the cql code.
> 
> A polymorphic view will be slower. In every operation involving the view, multiple virtual methods have to be called. With the variant, there is one virtual method call that dispatches to to generated code that is bound to the concrete view behind it, and so can be inlined.

Note that I'm not pushing for my polymorphic view idea, but I still want to clear that what I had in mind was a view with a single virtual method:
```c++
struct next_fragment {
    bytes_view data;
    bool has_more;
};
virtual next_fragment get_next() = 0;
```
This would mean that we have to call a single virtual per crossing fragment. 0 virtual calls is still better than 1 virtual call though, but I agree that we can sort this out once we have the major, most churn intensive changes in place.

avikivity commented:
> > > I have a slight preference towards a polymorphic view which in turn is a `FragmentedView` because I think it would be cleaner, but I want to hear more opinions as I'm not too familiar with the cql code.
> > 
> > 
> > A polymorphic view will be slower. In every operation involving the view, multiple virtual methods have to be called. With the variant, there is one virtual method call that dispatches to to generated code that is bound to the concrete view behind it, and so can be inlined.
> 
> Note that I'm not pushing for my polymorphic view idea, but I still want to clear that what I had in mind was a view with a single virtual method:
> 
> ```c++
> struct next_fragment {
>     bytes_view data;
>     bool has_more;
> };
> virtual next_fragment get_next() = 0;
> ```
> 
> This would mean that we have to call a single virtual per crossing fragment. 0 virtual calls is still better than 1 virtual call though, but I agree that we can sort this out once we have the major, most churn intensive changes in place.


It's actually 1 vs 1, if indeed it can be virtualized this way. The initial dispatch based on the type is a virtual call (a switch on the type is similar).

I thought we'd virtualize begin/end/etc. If we can do it in the way you suggest it would be equivalent, more or less. But things like copying an iterator, which are now done without thought, will suddenly involve an allocation and a virtual call.

michoecho commented:
I wrote a second possible version of raw_value_view. This time, instead of making a variant and visiting every time we want to do something with the data, I smashed the two view types together into a more general one, which remembers the underlying type and uses it to cross fragment boundaries. I pushed it here: https://github.com/michoecho/scylla/pull/9.

Expected difference: less code bloat, less lambdas to compile. I did no measurements though. More importantly, the interface is different. The first version requires adding deserialization/validation helpers to raw_value_view (or calling a visitor with a lambda every time we want to examine the value), while the second version is simply passed to generic FragmentedView functions.

The reason why raw_value_view and raw_value_view::view are separate types is because of that pesky shared pointer in raw_value_view. Because of it raw_value_view is not really a view. We don't want to copy that shared pointer around to validation and deserialization functions.

@denesb Take a look. Which version seems better?

avikivity commented:
> I wrote a second possible version of raw_value_view. This time, instead of making a variant

But raw_value_view is already a variant, no? You just added another branch?

Of course, if we have a path to de-variant it, that is good. But given it owns temporaries now, that will be difficult.

> and visiting every time we want to do something with the data, I smashed the two view types together into a more general one, which remembers the underlying type and uses it to cross fragment boundaries. I pushed it here: [michoecho#9](https://github.com/michoecho/scylla/pull/9).
> 
> Expected difference: less code bloat, less lambdas to compile. I did no measurements though. More importantly, the interface is different. The first version requires adding deserialization/validation helpers to raw_value_view (or calling a visitor with a lambda every time we want to examine the value), while the second version is simply passed to generic FragmentedView functions.
> 
> The reason why raw_value_view and raw_value_view::view are separate types is because of that pesky shared pointer in raw_value_view. Because of it raw_value_view is not really a view. We don't want to copy that shared pointer around to validation and deserialization functions.
> 
> @denesb Take a look. Which version seems better?

avikivity commented:
I commented on the alternative.

michoecho commented:
I added a second part to this series. See the updated PR message.
I went with the `std::variant` version of `raw_value_view` after all. It's easier to change to the other one, than vice versa.

michoecho commented:
Changing the title from `cql3: make cql3::raw_value fragmented` to `cql3: remove linearizations in the write path`.

michoecho commented:
cc @dekimir. Particularly the changes under `expr/` and `restrictions/`.

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -261,7 +261,7 @@ bool limits(const std::vector<column_value>& columns, const oper_t op, term& t,
         const auto cmp = get_value_comparator(columns[i])->compare(
                 // CQL dictates that columns[i] is a clustering column and non-null.
-                *get_value(columns[i], bag),
+                managed_bytes_view(*get_value(columns[i], bag)),
```
Why is this necessary?  Doesn't `get_value` return a `managed_bytes_view` now?

michoecho replied:
Right, it's not neccessary because the conversion from `managed_bytes` to `managed_bytes_view` can be implicit here. Removed.

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -322,7 +322,7 @@ bool contains(const column_value& col, const raw_value_view& value, const column
     const auto collection = get_value(col, bag);
     if (collection) {
-        return contains(col.col->type->deserialize(*collection), value);
+        return contains(col.col->type->deserialize(managed_bytes_view(*collection)), value);
```
Isn't `collection` already a `managed_bytes_view`?

michoecho replied:
No, `*collection` is a `managed_bytes`. `deserialize` is a template (which takes various views), so we need to make an explicit conversion to a view here.

avikivity replied:
It's sad to force this on users of abstract_type::desrialize, it should have a helper accepting a managed_bytes.

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -603,14 +611,32 @@ value_list get_IN_values(const ::shared_ptr<term>& t, size_t k, const query_opti
 static constexpr bool inclusive = true, exclusive = false;
 
+/// A range of all X such that X op val.
+nonwrapping_range<managed_bytes> to_range(oper_t op, managed_bytes&& val) {
```
Did you try modifying the template version (on line 678) to take a moved value?  That would be preferable to adding this near-identical overload.

michoecho replied:
Done.

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -689,12 +715,12 @@ value_set possible_lhs_values(const column_definition* cdef, const expression& e
                             }
                             if (is_compare(oper.op)) {
-                                const auto val = to_bytes_opt(oper.rhs->bind_and_get(options));
+                                managed_bytes_opt val = to_managed_bytes_opt(oper.rhs->bind_and_get(options));
```
Nitpick: why change `const auto`?  That change isn't strictly necessary, methinks.

michoecho replied:
Removing the `const` is needed to move `val` (into `to_range`). As for replacing `auto` with an explicit type, no particular reason.

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -734,26 +760,26 @@ value_set possible_lhs_values(const column_definition* cdef, const expression& e
                                 return unbounded_value_set;
                             }
-                            const auto val = to_bytes_opt(oper.rhs->bind_and_get(options));
+                            const auto val = to_managed_bytes_opt(oper.rhs->bind_and_get(options));
```
See, I _knew_ it wasn't necessary... :)

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -136,11 +136,10 @@ using value_set = std::variant<value_list, nonwrapping_range<bytes>>;
 /// A range of all X such that X op val.
-template<typename T>
-nonwrapping_range<T> to_range(oper_t op, const T& val);
+nonwrapping_range<clustering_key_prefix> to_range(oper_t op, const clustering_key_prefix& val);
```
This isn't quite what I meant.  Why couldn't we keep the declaration templated, just with a `T&&` parameter?  And the specialized `to_range<clustering_key_prefix>` declaration in expressions.cc wouldn't need a body.  A templated `to_range` is more widely useful than one that only takes a `clustering_key_prefix` parameter.

michoecho replied:
Template declarations are tricky to maintain, because you have to manually keep track of all uses and explicitly instantiate them all. If you don't, linking may succeed by accident (because an implicit instantiation happened to be generated), and then fail in a different build mode or a few patches later.
With forwarding references they are double tricky, because now `T` can also be a reference type. In other words, we would have to explicitly instantiate not only `to_range<clustering_key_prefix>`, but also `to_range<const clustering_key_prefix&>`, `to_range<clustering_key_prefix&>` and `to_range<clustering_key_prefix&&>`, depending on which of them are actually used. I think it's more reasonable to declare concrete versions in the header and implement them using the template.
Alternatively we could just move the implementation to the header.

dekimir replied:
> Template declarations are tricky to maintain, because you have to manually keep track of all uses and explicitly instantiate them all.

Yes, that's precisely what I was explicitly [asked](https://github.com/scylladb/scylla/pull/8227#discussion_r589007986) to do with `to_range`.  Perhaps @avikivity should chime in.

> Alternatively we could just move the implementation to the header.

That's how it started, but I was advised against it. :shrug:

avikivity replied:
I don't have the context here.

dekimir replied:
Context: this `to_range` function translates a binary_operator into an interval.  I templatized it so it can be used on multiple types.  You asked me to move the definition into the .cc file.  Now @michoecho wants to either detemplatize it or move the definition back into the header.

avikivity replied:
Well, de-templating is the best (as long as it's not replaced with manual duplication).

Second best is templates-but-not-in-the-header (as long as it doesn't involve long lists of explicit instantiations).

Third best is templates in the header.

dekimir replied:
> Well, de-templating is the best (as long as it's not replaced with manual duplication).

But it will be.  The very next thing I'll need is `to_range` for partition ranges.

dekimir replied:
To clarify my doubt: why would `foo(T&&)` ever instantiate into anything other than `foo(int&&)` for any invocation of foo on an integer reference?

michoecho replied:
That's because `T&&` is not an rvalue reference, but a forwarding reference.

Example:
```
int a;
const int b;

template <typename T> f(const T& x) {}
f(a); // T is deduced to int
f(b); // T is deduced to int
f(0); // T is deduced to int

// However...

template <typename T> g(T&& x) {}
g(a); // T is deduced to int&
g(b); // T is deduced to const int&
g(0); // T is deduced to int
```

See https://en.cppreference.com/w/cpp/utility/forward.

dekimir replied:
@michoecho since you got lots of work to do elsewhere in this PR, maybe it's best to leave this as it is now (ie, detemplatized).  It tops Avi's goodness scale when considered individually.  If/when the future arrives where I need multiple copies, I'll templatize it again.

michoecho replied:
Okay. But why would you need a public template, anyway? If you only need 2-3 versions of this function (I thought that was the case?), then isn't it the best two just write 2-3 declarations in the header, one template in the `.cc`, and implement those public declarations using the template, as I did in this patch? I don't see the downsides. There is some duplication, but we are only duplicating the declarations (and the one-liner bodies in .cc), not the logic.

dekimir replied:
> isn't it the best to just write 2-3 declarations in the header, one template in the `.cc`, and implement those public declarations using the template

No, that's second best. ;) It's strictly more typing than 1 header declaration and 2-3 bodyless instantiation declarations in the .cc file.  Both initially and upon every future change of function name or signature.

dekimir replied:
OTOH, it's easier to forget to instantiate something (especially with forward references, apparently) and cause a problem that only becomes visible two months later for someone else...

denesb commented on cql3/constants.hh:
```diff
@@ -73,7 +73,7 @@ public:
         virtual cql3::raw_value get(const query_options& options) override { return _bytes; }
         virtual cql3::raw_value_view bind_and_get(const query_options& options) override { return _bytes.to_view(); }
-        virtual sstring to_string() const override { return to_hex(*_bytes); }
+        virtual sstring to_string() const override { return _bytes.to_view().with_linearized([] (bytes_view bv) { return to_hex(bv); }); }
```
Can't we use `with_value` here to avoid a new call to `with_linearized?

michoecho replied:
We would need a generalized `to_hex`. Okay, I will do that.

michoecho replied:
Done.

avikivity commented on cql3/values.hh:
```diff
@@ -86,38 +96,67 @@ public:
+    ValueType deserialize(const collection_type_impl& t, cql_serialization_format sf) const {
+        return value_cast<ValueType>(with_value([&] (const FragmentedView auto& v) { return t.deserialize(v, sf); }));
+    }
+
```
This patch does two things: change the data structure, and change the API. This mixes an intense-but-small change with a boring-but-large change, requiring effort from the reviewer. Please split it into a patch that only changes the APIs but without any semantic change, and a patch that changes the data structure to be fragmented (and so all the semantic changes will be concentrated into that patch).

The patch that changes the API should explain the new APIs and the motivation for introducing them.

michoecho replied:
Done.

I would have done that from the start if only splitting header-touching commits wasn't so annoying (because of #1).

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -134,7 +134,8 @@ public:
+    template <FragmentedView View>
+    explicit managed_bytes(View v);
 
     managed_bytes(initialized_later, size_type size) {
```
This also belongs in its own patch.

michoecho replied:
Will do.

michoecho replied:
Done.

avikivity commented on types/tuple.hh:
```diff
@@ -141,6 +141,24 @@ public:
         return ret;
     }
+    template <typename RangeOf_bytes_opt>  // also accepts bytes_view_opt
+    static managed_bytes build_value_fragmented(RangeOf_bytes_opt&& range) {
```
Constraint?

michoecho replied:
This function only exists for a part of this series. It is later replaced with a version which takes fragmented input, and that one has a constraint.

avikivity commented on types/tuple.hh:
```diff
@@ -141,6 +141,24 @@ public:
     }
+    template <typename RangeOf_bytes_opt>  // also accepts bytes_view_opt
+    static managed_bytes build_value_fragmented(RangeOf_bytes_opt&& range) {
+        size_t size = 0;
```
I'm surprised it doesn't accept fragmented input, too. But I'm all for incremental development, so no hurry.

michoecho replied:
It is replaced with a version which takes fragmented input in one of the next patches.

avikivity replied:
Good.

avikivity commented on types/tuple.hh:
```diff
@@ -102,6 +102,24 @@ std::optional<View> read_tuple_element(View& v) {
+        if (v.empty()) {
+            return std::nullopt;
+        }
+        read_tuple_element(v);
```
Potential for optimization: skip_tuple_element, to avoid constructing and then destroying a bytes_opt. Let's defer it though.

michoecho replied:
`read_tuple_element` reads a view, not a `bytes_opt`, so it's unlikely to be an optimization.

avikivity replied:
It reads a view but returns a byte_opt, so it has to allocate storage for the return value.

michoecho replied:
No, `read_tuple_element` returns a view. I think you have misread it.

avikivity replied:
Yes.

avikivity commented on types/tuple.hh:
```diff
@@ -123,7 +141,18 @@ public:
         return _types;
     }
-    std::vector<bytes_view_opt> split(bytes_view v) const;
+    std::vector<bytes_opt> split(FragmentedView auto v) const {
```
This is a pessimization since we will now construct a bytes_opt for every element, possibly allocating storage. I think it's fine for now, since tuples are rare. We should revisit in the future and return a vector<fragmented view>.

michoecho replied:
This looks like a regression at a glance, but `split` was actually only used for two things:
1. Converting the returned `vector<bytes_view_opt>` to `vector<bytes_opt>`.
2. Getting the n-th element; I replaced this use with a separate function (`get_nth_tuple_element`).

avikivity replied:
Ok.

avikivity commented on types.cc:
```diff
@@ -703,6 +703,52 @@ void write_collection_value(bytes_ostream& out, cql_serialization_format sf, ato
+    }
+}
+
+template<typename T>
```
Constraint

michoecho replied:
Done.

avikivity commented on types/collection.hh:
```diff
@@ -54,6 +54,8 @@ public:
     template <typename BytesViewIterator>
     static bytes pack(BytesViewIterator start, BytesViewIterator finish, int elements, cql_serialization_format sf);
+    template <std::forward_iterator BytesViewIterator>
+    static managed_bytes pack_fragmented(BytesViewIterator start, BytesViewIterator finish, int elements, cql_serialization_format sf);
```
Please add a constraint on the value_type of the iterator.

michoecho replied:
This function is later replaced with a version which takes fragmented inputs, and that final version has the constraint, but I can add it to this commit too.

avikivity replied:
I see that in a later patch (6756eaa9df4829c8ad9ec4f6835e1f5cbee62258) even the std::forward_iterator is constraint is removed. Confused.

michoecho replied:
Oops, the change in 6756eaa doesn't do anything. It shouldn't be there, sorry for the confusion. e1ce9b0a add the constraints.

michoecho replied:
Fixed.

avikivity commented on types.cc:
```diff
@@ -749,6 +749,20 @@ void write_collection_value(managed_bytes_mutable_view& out, cql_serialization_f
     write_fragmented(out, single_fragmented_view(val));
 }
 
+void write_collection_value(managed_bytes_mutable_view& out, cql_serialization_format sf, const managed_bytes_view& val) {
```
I don't understand this. It isn't used in this patch, at least not directly. Shouldn't it be part of the previous patch?

michoecho replied:
It should. Moved to the previous patch. (It's only used later, so it doesn't really matter anyway though.)

avikivity commented on types/collection.hh:
```diff
@@ -123,9 +132,10 @@ collection_type_impl::pack(BytesViewIterator start, BytesViewIterator finish, in
 
-template <typename BytesViewIterator>
+template <typename Iterator>
+requires requires (Iterator it) { {*it} -> std::convertible_to<bytes_view>; }
```
Is it not a BytesViewIterator now? Do not understand the name change.

michoecho replied:
It is an iterator over either `bytes_view` or `bytes`, so the name wasn't great in the first place. The constraint expresses the intent better than the name.

avikivity replied:
Ok. We can introduce a named concept for this thing, but that can be done later.

avikivity commented on types/collection.hh:
```diff
@@ -139,3 +149,21 @@ collection_type_impl::pack_fragmented(BytesViewIterator start, BytesViewIterator
+    size_t len = collection_size_len(sf);
+    size_t psz = collection_value_len(sf);
+    for (auto j = start; j != finish; j++) {
+        len += managed_bytes_view(*j).size_bytes() + psz;
```
If we had size_bytes(managed_bytes_view) and size_bytes(bytes_view), perhaps the duplication here won't be necessary. But if you plan to eliminate the duplication soon, we can tolerate it.

michoecho replied:
Oh, 06f1f357 makes this function unused. I forgot to remove it.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -524,6 +524,10 @@ inline bool operator==(const managed_bytes_view& a, const managed_bytes_view& b)
 
+inline std::strong_ordering operator<=>(const managed_bytes_view& a, const managed_bytes_view& b) {
+    return compare_unsigned(a, b) <=> 0;
+}
```
Well, this is risky, since managed_bytes is used to hold all sorts of data and this <=> is not correct for all data types. So it will be easy to misuse in situations where we need to defer to the type's dynamic compare operation.

michoecho replied:
You could say the same for `bytes`, though, but that kind of comparison for `bytes` is in use. This change was needed to replace `nonwrapping_range<bytes>` with `nonwrapping_range<managed_bytes>` in `expression.cc`.

dekimir replied:
That's unexpected, as expression.cc is supposed to use the data_type comparator, not byte operators.

avikivity replied:
It's true that managed_bytes is both a container for bytes and an container for data that has to be interpreted by the type system.

 I don't see why nonwrapping_range requires operator<=>, it is explicitly for use with a supplied comparator. Perhaps it indicates a bug in nonwrapping_range or expression.cc.

michoecho replied:
Hmm, I will check again where that operator<=> is used.

dekimir replied:
I would recommend against having a comparison operator on ranges; it's almost guaranteed to confuse some users into thinking they can check for overlap with it.  But that's not true (see, eg, #8157).

michoecho replied:
Okay, I did something weird. This operator really isn't used. I will remove it.
I would swear I wrote it because something wasn't compiling without it. I don't know what happened.

michoecho replied:
Fixed.

michoecho commented:
v2:
- Split the commit fragmenting `raw_value` into three commits: one which add a more general API (which doesn't expose the internals), one which switches all users to that API, and one that changes the internals.
- Similarly split other commits where applicable (but not all: for examle `cql3: don't linearize elements of lists, tuples, and user types` still is fairly chaotic, but I think that splitting it would require disproportionate effort)
- Fixed the confusing constraints on `pack_fragmented` in the middle commits.
- Addressed all the other minor change requests.

avikivity commented on utils/fragment_range.hh:
```diff
@@ -410,3 +410,23 @@ void write_native(Out& out, std::type_identity_t<T> v) {
     }
 }
+
+static inline sstring::iterator fragment_to_hex(sstring::iterator out, bytes_view frag) {
```
`static inline` shouldn't be used in C++ headers. It causes each translation unit to instantiate the function separately, and if the compiler chooses not to inline, we end up with one copy per translation unit that instantiates the function.

In addition, it shouldn't be in a header at all (see #1). We shouldn't have headers, but that will take more work.

avikivity replied:
(plain `inline` allows the linker to merge duplicate instantiations)

avikivity commented on cql3/values.hh:
```diff
@@ -117,6 +121,56 @@ public:
+    ValueType deserialize(const abstract_type& t) const {
+        return value_cast<ValueType>(with_value([&] (const FragmentedView auto& v) { return t.deserialize(v); }));
+    }
+
```
Minor: we call abstract_type::deserialize which arranges for a high-level data_value to be produced, complete with a lw_shared_ptr for the type, then throw the type away. We should have a lower-level entry point that allows us to avoid that useless work. It can be done later.

michoecho replied:
I agree. It a bit insane that there's a memory allocation involved to read an integer from memory.

avikivity commented on cql3/attributes.cc:
```diff
@@ -123,7 +122,7 @@ db::timeout_clock::duration attributes::get_timeout(const query_options& options
         throw exceptions::invalid_request_exception("Timeout value cannot be unset/null");
     }
-    cql_duration duration = value_cast<cql_duration>(duration_type->deserialize(*timeout));
+    cql_duration duration = timeout.deserialize<cql_duration>(*duration_type);
```
There is redundancy in supplying both the run-time type and the compile-time type. It is both inefficient, and can blow up if we supply conflicting types. Not suggesting to do anything about now, but keep it in mind for future improvement.

avikivity commented on cql3/update_parameters.hh:
```diff
@@ -157,6 +157,22 @@ public:
+        });
+    };
+
     atomic_cell make_cell(const abstract_type& type, const fragmented_temporary_buffer::view& value, atomic_cell::collection_member cm = atomic_cell::collection_member::no) const {
```
Can we drop this variant now?

avikivity commented on test/boost/transport_test.cc:
```diff
@@ -26,6 +26,22 @@
+        return a.is_null() == b.is_null();
+    }
+}
+
```
The commit message says it's moved, but it's also rewritten. Why is that? did we lose access to some private members?

michoecho replied:
The old version both uses private members (and we wanted to move this operator out of the class), and uses `operator*` which is removed in the next patch.

avikivity commented on types.cc:
```diff
@@ -708,6 +708,66 @@ void write_collection_value(bytes_ostream& out, cql_serialization_format sf, ato
+}
+
+template<std::integral T>
+void write_simple(managed_bytes_mutable_view& out, std::type_identity_t<T> val) {
```
Nice trick with type_identity_t.

avikivity commented on types.cc:
```diff
@@ -708,6 +708,66 @@ void write_collection_value(bytes_ostream& out, cql_serialization_format sf, ato
+        auto p = out.current_fragment().data();
+        out.remove_prefix(sizeof(T));
+        // FIXME use write_unaligned after it's merged.
+        memcpy(p, &val, sizeof(T));
```
The compiler will optimize the memcpy to a single instruction just fine.

avikivity commented on types.cc:
```diff
@@ -708,6 +708,66 @@ void write_collection_value(bytes_ostream& out, cql_serialization_format sf, ato
+        write_simple<uint16_t>(out, uint16_t(val.size_bytes()));
+    }
+    write_fragmented(out, val);
+}
```
(and collections certainly don't need to be hyper-optimized).

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -185,6 +193,18 @@ bool equal(const bytes_opt& rhs, const column_value& lhs, const column_value_eva
 }
 
+/// True iff lhs's value equals rhs.
+bool equal(const managed_bytes_opt& rhs, const column_value& lhs, const column_value_eval_bag& bag) {
```
I love expression.cc's convention of putting rhs on the left, and lhs on the right.

avikivity commented on cql3/update_parameters.hh:
```diff
@@ -187,20 +187,6 @@ public:
-            return atomic_cell::make_live(type, _timestamp, value, cm);
-        }
-    };
-
```
I see it got eventually removed.

avikivity commented:
Very good. I left a few minor/trivial comments. Please run debug-mode unit tests, and release-mode `perf_simple_query --smp 1` and add the before/after results to the cover letter.

michoecho commented:
`perf_simple_query --smp 1` results (every value shown is a median of a 10 second run, and the test was repeated 5 times):

```
old       new       %diff

read:
136935.80 135470.36 -1.07017
137548.62 135090.89 -1.78681
137712.60 134804.71 -2.11156
134618.11 134765.97 0.109837
136954.88 134700.67 -1.64595

write:
149522.70 149809.05 0.191509
149085.69 148517.55 -0.381083
149731.40 150223.83 0.328876
149141.60 146767.93 -1.59155
149920.03 150064.19 0.0961579
```
So there is almost no difference for `--write`, but a visible slowdown for reads. I will try the version with a `bytes_view` variant added too.

michoecho commented:
v3:
- removed `static` from `static inline`
- rebased

avikivity commented:
@michoecho sometimes a change can cause a 1-2% performance difference for no reason, just because the functions fit better in the cache due to some random code generation change. Here it's probably real, but I'm willing to take a 1-2% hit for this. But I'll wait for a report with the single bytes_view path.

michoecho commented:
New `perf_simple_query --smp 1` results:
```
old       new       %wrt old bytesview %wrt old

read
153074.70 148251.58 -3.15083 150011.33 -2.00123
149291.98 148414.95 -0.58746 151749.77 +1.64630
152350.54 150323.07 -1.33079 153951.08 +1.05056
153575.42 150623.14 -1.92236 152646.21 -0.60505
153903.93 149886.32 -2.61047 153105.04 -0.51908
write
163180.56 161714.64 -0.89834 159150.43 -2.46974
161499.80 163102.08 +0.99212 163126.01 +1.00694
164011.46 165197.26 +0.72299 162748.56 -0.77000
163940.56 165778.94 +1.12137 164248.33 +0.18773
163227.15 165113.54 +1.15568 164231.42 +0.61525
```
I don't know what to think about this. It seems that it does slightly better on read, and slightly worse on write, but maybe it's really all just random.
Note: the absolute results are different this time because apparently my CPU was clocking at 4.0Ghz last time for reasons I don't know. The new results are fresh after a reboot, at full 4.5Ghz.

avikivity commented:
It could be just random. Is this with or without the dedicated bytes_view path?

I don't like dismissing bad results as "just random" since it's an easy way to accumulate small regressions, but:
1. the linearizations, when they happen, are far worse, and 1-2% is an acceptable price
2. we don't have much of a choice, short of building a micro-benchmark for this to emphasize the areas the code changed
3. it could in fact really be just random

michoecho commented:
>Is this with or without the dedicated bytes_view path

There are 3 columns in the table. `old` is before this series, `new` is after this series, `bytesview` is after this series, with the dedicated `bytes_view` path added.

avikivity commented:
Well, that's a puzzler. Pick the one you like and I'll merge it.

michoecho commented:
If there is no clear reason to add a special case, then it makes sense to pick the version without the special case. So I pick the version in this PR.

avikivity commented:
Queued.
